### ISSUE TITLE
function getSearchContext in extension of modeladmin is not called anymore 

### ISSUE THREAD

**Issue by lienajp** [silverstripe/silverstripe-framework#8827]
## Affected Version 
v4.3.0

## Description
When extending ModelAdmin to add new pre-defined queries (e.g. a checkbox that will be used to filter adresses in Belgium - using later on the getlist function), the overwritten getSearchContext functions never called.
This is the code I used but you may simply use examples from your on-line developers doc (https://docs.silverstripe.org/en/4/developer_guides/customising_the_admin_interface/modeladmin/).
 
```php
class Adress_Admin extends ModelAdmin {
    private static $url_segment = 'Adress';
    private static $menu_title = 'Adressen';
    private static $managed_models = [Adress::class];

    public function getSearchContext() 
    {
        $context = parent::getSearchContext();
        if($this->modelClass == 'bpf\Adress') {
            $context->getFields()->push(new CheckboxField('q[InBelgiumA]', 'Only expensive stuff'));
        }
        return $context;
    }
```

## Steps to Reproduce
It's very easy to check - use e.g. error_log to trace the fact the function has been called.

Help us with step-by-step instructions.

ps: I know this will be depreciated in v5 but at least could you please provide an updated documentation on how to do it using GridfieldFilterHeader.
Thank you!!!

**Comment by ScopeyNZ** [comment]
We made an effort to ensure the existing API still worked - apparently we missed a spot. 

The extension points for updating context still exist. You can write an extension that uses the `updateSearchContext` extension point:

https://github.com/silverstripe/silverstripe-admin/blob/8267623bc5044e47026d7716e859d2208b0e651e/code/ModelAdmin.php#L201

**Comment by lienajp** [comment]
Hi Guy,

It looks a bit complex to me but I will try to do it.
I learned silverstripe all by myself without little support and I get lost sometimes. My actual site in Prod runs with V3.5.1 and I am working on converting it to V4.3.
Not so easy as some facilities/modules became obsolete e.g. Translatable.

ok - this being said, what I want to do is what is explained on the developers documentation. In casus the following:
It would be fine if you can show the equivalent with GridfieldFilterHeader as this is the future no?

Thank you in advance
Philippe
 
You can also customize the search behavior directly on your ModelAdmin instance. For example, we might want to have a checkbox which limits search results to expensive products (over $100).

app/code/MyAdmin.php

use SilverStripe\Forms\CheckboxField;
use SilverStripe\Admin\ModelAdmin;

class MyAdmin extends ModelAdmin 
{
    public function getSearchContext() 
    {
        $context = parent::getSearchContext();

        if($this->modelClass == 'Product') {
            $context->getFields()->push(new CheckboxField('q[ExpensiveOnly]', 'Only expensive stuff'));
        }

        return $context;
    }

    public function getList() 
    {
        $list = parent::getList();

        $params = $this->getRequest()->requestVar('q'); // use this to access search parameters

        if($this->modelClass == 'Product' && isset($params['ExpensiveOnly']) && $params['ExpensiveOnly']) {
            $list = $list->exclude('Price:LessThan', '100');
        }

        return $list;
    }
}


On 26 Feb 2019, at 22:37, Guy Marriott <notifications@github.com> wrote:

We made an effort to ensure the existing API still worked - apparently we missed a spot.

The extension points for updating context still exist. You can write an extension that uses the updateSearchContext extension point:

https://github.com/silverstripe/silverstripe-admin/blob/8267623bc5044e47026d7716e859d2208b0e651e/code/ModelAdmin.php#L201 <https://github.com/silverstripe/silverstripe-admin/blob/8267623bc5044e47026d7716e859d2208b0e651e/code/ModelAdmin.php#L201>

**Comment by lienajp** [comment]
I succeeded to create the extension and get it called. The additional variable is added to the form.
However, it looks like the parameters are not passed using the URL anymore, which means that <code> $params = $this->getRequest()->requestVar('q‘) </code> returns nothing.
I don’t know how to get the value for the additional variable I created. 

In app.yml, I added

SilverStripe\Admin\ModelAdmin:
  extensions:
    - bpf\ModelAdminExtension

I created the class ModelAdminExtension

<?php    
namespace bpf;

use SilverStripe\Admin\ModelAdmin;
use SilverStripe\i18n\i18nEntityProvider;
use SilverStripe\Forms\CheckboxField;
use SilverStripe\Forms\Form;
use SilverStripe\Forms\FormAction;
use SilverStripe\Forms\GridField\GridFieldDetailForm;
use SilverStripe\Forms\GridField\GridFieldFilterHeader;
use SilverStripe\ORM\DataExtension;

class ModelAdminExtension extends DataExtension
{
    public function updateSearchContext($context)
    {
        if($this->owner->modelClass == Adress::class) {
            $context->getFields()->push(new CheckboxField('q[InBelgiumA]', 'Only if in Belgium'));
        }
        return $context;
    }
}
?>

I do not get any params in the getlist() function of my  modeladmin extension because params are not passed using the URL.


class Adress_Admin extends ModelAdmin {
    private static $url_segment = 'Adress';
    private static $menu_title = 'Adressen';
    private static $managed_models = [Adress::class];
    private static $menu_icon_class = 'font-icon-book';

    public function getList() 
    {
        $list = parent::getList();
        $params = $this->getRequest()->requestVar('q'); // use this to access search parameters

        if($this->modelClass == Adress::class && isset($params['InBelgiumA']) && $params['InBelgiumA']) {
            $list = $list->exclude('getInBelgium()', true);
        }

        return $list;
    }
}


On 26 Feb 2019, at 22:37, Guy Marriott <notifications@github.com> wrote:

We made an effort to ensure the existing API still worked - apparently we missed a spot.

The extension points for updating context still exist. You can write an extension that uses the updateSearchContext extension point:

https://github.com/silverstripe/silverstripe-admin/blob/8267623bc5044e47026d7716e859d2208b0e651e/code/ModelAdmin.php#L201 <https://github.com/silverstripe/silverstripe-admin/blob/8267623bc5044e47026d7716e859d2208b0e651e/code/ModelAdmin.php#L201>

**Comment by Leapfrognz** [comment]
Yes, @ScopeyNZ The example for customising ModelAdmin search as shown in the docs doesnt work anymore and is throwing a lot of people. We need an example in there using the preferred method with `GridfieldFilterHeader`.

**Comment by anhld** [comment]
With SS 4.3.x and upward,  the [ModelAdmin custom search box guide](https://docs.silverstripe.org/en/4/developer_guides/customising_the_admin_interface/modeladmin/#displaying-results) is not valid anymore. I've figured out a way to get the job done.

1. Use `updateSearchContext` to extend search form:

```php
public function updateSearchContext($context)
{
    // don't need to wrap field name inside the `q[]` tag anymore
    $dateField = DateField::create('DateFrom', 'From date');

    $context->getFields()->push($field);
}
```

2. Add following function to your custom ModelAdmin:

```php
/**
 * Returns GridField filter params
 */
public function getFilterParams()
{
    $class = $this->sanitiseClassName($this->modelClass);
    $body  = $this->getRequest()->getBody();

    parse_str($body, $array);

    $params = [];

    // first, try to get params from filter state in request body
    if (isset($array['filter'][$class]) && is_array($array['filter'][$class])) {
        $params = $array['filter'][$class];
    }

    // second, try to get params from GridFieldFilterHeader state in request body
    if (empty($params) && isset($array[$class]['GridState']) && $array[$class]['GridState']) {
        $gridState = json_decode($array[$class]['GridState'], true);

        if (isset($gridState['GridFieldFilterHeader']['Columns'])
            && is_array($gridState['GridFieldFilterHeader']['Columns'])) {
            $params = $gridState['GridFieldFilterHeader']['Columns'];
        }
    }

    // third, try to get params from GridFieldFilterHeader state in request vars
    if (empty($params) && $requestVar = $this->getRequest()->requestVar($class)) {
        if (isset($requestVar['GridState']) && !empty($requestVar['GridState'])) {
            $gridState = json_decode($requestVar['GridState'], true);

            if (isset($gridState['GridFieldFilterHeader']['Columns'])
                && is_array($gridState['GridFieldFilterHeader']['Columns'])) {
                $params = $gridState['GridFieldFilterHeader']['Columns'];
            }
        }
    }

    return $params;
}
```

3. Use `getFilterParams` to get search params and do filter:

```php
public function getList()
{
    // get list
    $list = parent::getList();

    // get search params
    $params = $this->getFilterParams();

    if (isset($params['DateFrom'])) {
        // do filter stuffs
    }
    
    return $list;
}
```

**Comment by lienajp** [comment]
Great!!! Thank you Anh Le!!!
This works fine if you correct the typo in the function updateSearchContext:
`
public function updateSearchContext($context)
{
    // don't need to wrap field name inside the `q[]` tag anymore
    $dateField = DateField::create('DateFrom', 'From date');

    // $context->getFields()->push($field); this should be ... next line
    $context->getFields()->push($dateField);
}
`

**Comment by ScopeyNZ** [comment]
I'm going to reopen this in the context of docs because there's a lot of good discussion here.

**Comment by Jellygnite** [comment]
I added this to my ModelAdmin but the updateSearchContext function doesn't seem to even be called. Am I missing something from @anhld  solution? Running SS4.3.2. Any help would be greatly appreciated. 
@lienajp , you said you got this working...?
```
class JobAdmin extends ModelAdmin
{
	private static $managed_models = [
		'Job'
	];
	public function updateSearchContext($context)
	{
		// don't need to wrap field name inside the q[] tag anymore
		$dateField = DateField::create('DateFrom', 'From date');
		$context->getFields()->push($dateField);
	}
}
```

**Comment by Jellygnite** [comment]
Forgive my ignorance. I didn't realise you had to extend your custom ModelAdmin. For those looking for a more complete solution:
https://docs.silverstripe.org/en/4/developer_guides/customising_the_admin_interface/how_tos/extending_an_existing_modeladmin/

```
use SilverStripe\Core\Extension;
use SilverStripe\Forms\DateField;

class JobAdminExtension extends Extension 
{
	
	public function updateSearchContext($context)
	{
		// don't need to wrap field name inside the `q[]` tag anymore
		$dateField = DateField::create('DateFrom', 'From date');
		$context->getFields()->push($dateField);
	}
}
```

And in your config.yml

```
JobAdmin:
  extensions:
    - JobAdminExtension
```

Working now : )

**Comment by Leapfrognz** [comment]
So you need an extension to get this working on a custom modelAdmin? That feels a bit of a roundabout way....

**Comment by Jellygnite** [comment]
Yes, that is why I didn't cotton on to it in the first place. It does seem a roundabout way of doing this.

**Comment by lienajp** [comment]
@Jellygnite: yes, I got it working. The extension thing is mentioned in my second reaction on 27/2. See also the title of this issue. You noticed quickly and managed to get it working, so that's fine.
It is indeed a workaround as the updateSearchContext function should be called. This is why it its considered as a bug (I guess).

**Comment by jonom** [comment]
Should the impact of this bug be higher? I've developed some pretty complex search forms for some clients because the default ones aren't very powerful, so losing those has a major impact for my users.

**Comment by jonom** [comment]
This is a related issue: https://github.com/silverstripe/silverstripe-framework/issues/8518 the quick fix there was not to fix this bug, but to allow opting out of the React search form so the bug isn't surfaced. To test that solution I tried setting `useLegacyFilterHeader` to for a single model admin instance by accessing the GridFieldFilterHeader instance from `getEditForm()`, but it didn't work, perhaps because it was switched after initialisation? I then tried forcing the legacy behaviour with:
```yml
SilverStripe\Forms\GridField\GridFieldFilterHeader:
  force_legacy: true
```
That triggered the 'legacy' behaviour, but apparently that behaviour is **not** equivalent to the behaviour in SS 4.2. In SS 4.2 when you clicked the search icon, a vertical search form expanded, showing all the fields you provided in `getSearchContext()`. This allowed you to create whatever fields and behaviour you like, which was great. (For example, in this interface the Community and Lot Status each search across a couple of different fields)

_4.2 model admin filter UI:_
![Screen Shot 2019-06-13 at 11 38 20 am](https://user-images.githubusercontent.com/1079425/59458830-2a114680-8dd0-11e9-97b5-5b2754df12c5.png)

The 'legacy' behaviour instead places search fields directly in columns, which means that you can only provide search fields that match up with columns 1:1. The bigger issue though, is that `getSearchContext()` is still not called, so my custom search form is not used whether I use the new React UI or the supposed 'legacy' UI. Instead the search fields are automatically scaffolded.

_4.3 model admin 'legacy' filter UI:_
![Screen Shot 2019-06-13 at 11 42 01 am](https://user-images.githubusercontent.com/1079425/59458988-8c6a4700-8dd0-11e9-9925-d3e36d376edc.png)

The solution outlined above by @anhld looks workable but I'm reluctant to do it since a refactor like that shouldn't be necessary after upgrading to a new minor version. I might just revert to 4.2 instead.

**Comment by lienajp** [comment]
I confirm the solution works and I fully agree with you that this is a bad surprise after a minor upgrade. 
But as I am stem in the process to migrate from 3.6 to 4.3 it was a satisfactory solution for me.

**Comment by michalkleiner** [comment]
Based on the posts above I'd agree that the impact should be at least medium.

**Comment by Leapfrognz** [comment]
> Should the impact of this bug be higher? I've developed some pretty complex search forms for some clients because the default ones aren't very powerful, so losing those has a major impact for my users.
I'm in the same basket with Jonom. We have some pretty complex CMS search forms that clients absolutely need.

**Comment by gheggie** [comment]
> 
> 
> So you need an extension to get this working on a custom modelAdmin? That feels a bit of a roundabout way....

In 4.3.3 you don't need an extension (not sure about other versions of 4.3.x though).

In the custom ModelAdmin, create the `getEditForm()` function and add this:

```
    $config->removeComponentsByType(GridFieldFilterHeader::class);
    $filterHeader = new GridFieldFilterHeader(
            false,
            function($context) {
                $context->getFields()->insertAfter(
                    'SomeField',
                    CheckboxField::create('NewField', 'My new field to filter on')
                );
            }
        );
    $config->addComponent($filterHeader);
```

**Comment by elkebe** [comment]
Got this to work, however if you pre-filter the list to exclude eg. 'Status' => 'completed' by default, when selected to include 'completed' items, they can't be opened in the GridField Detail Form... this may be a related bug?

**Comment by LakeeshaRCL** [comment]
> With SS 4.3.x and upward, the [ModelAdmin custom search box guide](https://docs.silverstripe.org/en/4/developer_guides/customising_the_admin_interface/modeladmin/#displaying-results) is not valid anymore. I've figured out a way to get the job done.
> 
> 1. Use `updateSearchContext` to extend search form:
> 
> ```
> public function updateSearchContext($context)
> {
>     // don't need to wrap field name inside the `q[]` tag anymore
>     $dateField = DateField::create('DateFrom', 'From date');
> 
>     $context->getFields()->push($field);
> }
> ```
> 
> 1. Add following function to your custom ModelAdmin:
> 
> ```
> /**
>  * Returns GridField filter params
>  */
> public function getFilterParams()
> {
>     $class = $this->sanitiseClassName($this->modelClass);
>     $body  = $this->getRequest()->getBody();
> 
>     parse_str($body, $array);
> 
>     $params = [];
> 
>     // first, try to get params from filter state in request body
>     if (isset($array['filter'][$class]) && is_array($array['filter'][$class])) {
>         $params = $array['filter'][$class];
>     }
> 
>     // second, try to get params from GridFieldFilterHeader state in request body
>     if (empty($params) && isset($array[$class]['GridState']) && $array[$class]['GridState']) {
>         $gridState = json_decode($array[$class]['GridState'], true);
> 
>         if (isset($gridState['GridFieldFilterHeader']['Columns'])
>             && is_array($gridState['GridFieldFilterHeader']['Columns'])) {
>             $params = $gridState['GridFieldFilterHeader']['Columns'];
>         }
>     }
> 
>     // third, try to get params from GridFieldFilterHeader state in request vars
>     if (empty($params) && $requestVar = $this->getRequest()->requestVar($class)) {
>         if (isset($requestVar['GridState']) && !empty($requestVar['GridState'])) {
>             $gridState = json_decode($requestVar['GridState'], true);
> 
>             if (isset($gridState['GridFieldFilterHeader']['Columns'])
>                 && is_array($gridState['GridFieldFilterHeader']['Columns'])) {
>                 $params = $gridState['GridFieldFilterHeader']['Columns'];
>             }
>         }
>     }
> 
>     return $params;
> }
> ```
> 
> 1. Use `getFilterParams` to get search params and do filter:
> 
> ```
> public function getList()
> {
>     // get list
>     $list = parent::getList();
> 
>     // get search params
>     $params = $this->getFilterParams();
> 
>     if (isset($params['DateFrom'])) {
>         // do filter stuffs
>     }
>     
>     return $list;
> }
> ```



> With SS 4.3.x and upward, the [ModelAdmin custom search box guide](https://docs.silverstripe.org/en/4/developer_guides/customising_the_admin_interface/modeladmin/#displaying-results) is not valid anymore. I've figured out a way to get the job done.
> 
> 1. Use `updateSearchContext` to extend search form:
> 
> ```
> public function updateSearchContext($context)
> {
>     // don't need to wrap field name inside the `q[]` tag anymore
>     $dateField = DateField::create('DateFrom', 'From date');
> 
>     $context->getFields()->push($field);
> }
> ```
> 
> 1. Add following function to your custom ModelAdmin:
> 
> ```
> /**
>  * Returns GridField filter params
>  */
> public function getFilterParams()
> {
>     $class = $this->sanitiseClassName($this->modelClass);
>     $body  = $this->getRequest()->getBody();
> 
>     parse_str($body, $array);
> 
>     $params = [];
> 
>     // first, try to get params from filter state in request body
>     if (isset($array['filter'][$class]) && is_array($array['filter'][$class])) {
>         $params = $array['filter'][$class];
>     }
> 
>     // second, try to get params from GridFieldFilterHeader state in request body
>     if (empty($params) && isset($array[$class]['GridState']) && $array[$class]['GridState']) {
>         $gridState = json_decode($array[$class]['GridState'], true);
> 
>         if (isset($gridState['GridFieldFilterHeader']['Columns'])
>             && is_array($gridState['GridFieldFilterHeader']['Columns'])) {
>             $params = $gridState['GridFieldFilterHeader']['Columns'];
>         }
>     }
> 
>     // third, try to get params from GridFieldFilterHeader state in request vars
>     if (empty($params) && $requestVar = $this->getRequest()->requestVar($class)) {
>         if (isset($requestVar['GridState']) && !empty($requestVar['GridState'])) {
>             $gridState = json_decode($requestVar['GridState'], true);
> 
>             if (isset($gridState['GridFieldFilterHeader']['Columns'])
>                 && is_array($gridState['GridFieldFilterHeader']['Columns'])) {
>                 $params = $gridState['GridFieldFilterHeader']['Columns'];
>             }
>         }
>     }
> 
>     return $params;
> }
> ```
> 
> 1. Use `getFilterParams` to get search params and do filter:
> 
> ```
> public function getList()
> {
>     // get list
>     $list = parent::getList();
> 
>     // get search params
>     $params = $this->getFilterParams();
> 
>     if (isset($params['DateFrom'])) {
>         // do filter stuffs
>     }
>     
>     return $list;
> }
> ```

I am quite new to SilverStripe. However, I migrated a SilverStripe project from 4.1.2 to 4.7.0. In the current project lot of custom search fields used with the getSearchContext() method. The Answer mentioned by @anhld looks to be correct but I need some help to do the modification in my project.

In my project, there were files extended by SearchContext, in those files search fields and queries were defined.  Also, several admin classes were defined using ModelAdmin (eg: UserSearchAdmin extends ModelAdmin). Inside those admin classes SearchContext objects were created inside getSearchContext() method. 

If anyone can mention to me where to add the updateSearchContext() method mentioned by @anhld it would be a great help for me.

Thank you.

**Comment by LakeeshaRCL** [comment]
I got it.
